### ISSUE TITLE
[data2vec] Loss goes down and up again

### ISSUE THREAD

**Issue by tomresan** [facebookresearch/fairseq#4177]
I try to train data2vec on music data (the FMA dataset). I've made some modifications to the feature extractor ConvNet (I've made it a small ResNet essentially), and reduced the size of the transformer encoder (8 layers, 8 attention heads, 512 emb dim, 2048 emb_ffn_dim). Since I am new to training transformer models, I've tried to reproduce the paper's hyperparameters as closely as possible. My batch size is much smaller, so I've adjusted the learning rate accordingly etc..

The first few epochs seem promising. The loss goes down quite quickly and the embedding variance of the teacher and student models increase over time respectively (i.e. the model is not collapsing to a constant representation). However, after a while it abruptly changes and the loss goes up again and stays high over the remaining training (the variances stay high though). I've also experienced similar issues with another exponential moving average (EMA) model, DINO (the ResNet variant), a while ago.

When reducing the learning rate, the effect just happens later.

There indirectly seems to be a similar issue [here](https://github.com/pytorch/fairseq/issues/4159#issue-1117209267).

Does anyone have an idea why this might be happening?

**Comment by tomresan** [comment]
Okay, so I've plotted the teacher and student variances from two test runs with a small dataset and only for a few epochs:

https://imgur.com/a/dynF7P9

The student variance begins to converge to the teacher variance until a specific point. After that the loss quickly increases and there is a relatively constant gab between the two variances.

Not sure how to interpret this. Maybe this is the desired behaviour, since the teacher should always be "better" in order to lead the student.

**Comment by Ramlinbird** [comment]
> 



> Okay, so I've plotted the teacher and student variances from two test runs with a small dataset and only for a few epochs:
> 
> https://imgur.com/a/dynF7P9
> 
> The student variance begins to converge to the teacher variance until a specific point. After that the loss quickly increases and there is a relatively constant gab between the two variances.
> 
> Not sure how to interpret this. Maybe this is the desired behaviour, since the teacher should always be "better" in order to lead the student.

Hello, @Tomsen1410 ! Have you found a reasonable explanation? Follow is my strange loss plot...
![image](https://user-images.githubusercontent.com/11084132/153812577-47635b23-4642-4084-a804-b942020a3f49.png)

**Comment by tomresan** [comment]
> > 
> 
> > Okay, so I've plotted the teacher and student variances from two test runs with a small dataset and only for a few epochs:
> > https://imgur.com/a/dynF7P9
> > The student variance begins to converge to the teacher variance until a specific point. After that the loss quickly increases and there is a relatively constant gab between the two variances.
> > Not sure how to interpret this. Maybe this is the desired behaviour, since the teacher should always be "better" in order to lead the student.
> 
> Hello, @Tomsen1410 ! Have you found a reasonable explanation? Follow is my strange loss plot... ![image](https://user-images.githubusercontent.com/11084132/153812577-47635b23-4642-4084-a804-b942020a3f49.png)

Hey!
Unfortunately not. Your loss seems to be ok though? I am not sure how the "correct" loss function should look like. It is strange that my loss values are two orders of magnitude smaller. Would also be great to see a training plot from the authors. Could you tell me which hyperparameters you have used? (batch size, ...)

**Comment by Ramlinbird** [comment]
> > > 
> > 
> > 
> > > Okay, so I've plotted the teacher and student variances from two test runs with a small dataset and only for a few epochs:
> > > https://imgur.com/a/dynF7P9
> > > The student variance begins to converge to the teacher variance until a specific point. After that the loss quickly increases and there is a relatively constant gab between the two variances.
> > > Not sure how to interpret this. Maybe this is the desired behaviour, since the teacher should always be "better" in order to lead the student.
> > 
> > 
> > Hello, @Tomsen1410 ! Have you found a reasonable explanation? Follow is my strange loss plot... ![image](https://user-images.githubusercontent.com/11084132/153812577-47635b23-4642-4084-a804-b942020a3f49.png)
> 
> Hey! Unfortunately not. Your loss seems to be ok though? I am not sure how the "correct" loss function should look like. It is strange that my loss values are two orders of magnitude smaller. Would also be great to see a training plot from the authors. Could you tell me which hyperparameters you have used? (batch size, ...)

Thanks. I didn't change any hyperparameters except `max_tokens` (since the memory error) in base_librispeech.yaml.
@alexeib could you share your training plot with us, and help us to figure out this? Thanks so much.

**Comment by alexeib** [comment]
hey, if variances are jumping up and down, that looks like a collapse and you may want to lower you learning rate. i dont have variance plots for the nlp models (wasnt logging it when it was trained) but here is the loss curve:

<img width="1294" alt="image" src="https://user-images.githubusercontent.com/3497946/154182962-414ea6dd-8264-429e-b357-c69c77aaa69c.png">

here are a couple example plots from a reduced speech setup with variance etc (it should look somewhat similar with nlp but not exactly the same). this is also using a tri-stage lr scheduler that holds learning rate at peak rate for 90% of the training:

loss:
<img width="1196" alt="image" src="https://user-images.githubusercontent.com/3497946/154183188-acd81628-77e1-4dba-8438-a874e262f8e7.png">

pred var:
<img width="1203" alt="image" src="https://user-images.githubusercontent.com/3497946/154183229-7bb26431-76c6-4c1b-8399-1c312c0bd84b.png">

target var:
<img width="1208" alt="image" src="https://user-images.githubusercontent.com/3497946/154183275-06352bfe-2ff8-44b9-8baf-0811bb8a62c4.png">

**Comment by Ramlinbird** [comment]
@alexeib Thanks a lot for your quick reply! According to your loss plot, my training seems all right? And I also check my variance plot, 
![image](https://user-images.githubusercontent.com/11084132/154185164-0c62f14c-e3de-4ec8-903b-2bf7c3a9f9aa.png)
However, the pred var and target var is not flat in the end, they are still dropping, is this OK? What's the actual meaning of these values, and what relation should they have? (I read the source code, they are just standard variance of network's outputs.)

**Comment by tomresan** [comment]
Was there also instance_norm applied to the targets at the reduced speech setup @alexeib ?

**Comment by Orlllem** [comment]
> hey, if variances are jumping up and down, that looks like a collapse and you may want to lower you learning rate. i dont have variance plots for the nlp models (wasnt logging it when it was trained) but here is the loss curve:
> 
> <img alt="image" width="1294" src="https://user-images.githubusercontent.com/3497946/154182962-414ea6dd-8264-429e-b357-c69c77aaa69c.png">
> 
> here are a couple example plots from a reduced speech setup with variance etc (it should look somewhat similar with nlp but not exactly the same). this is also using a tri-stage lr scheduler that holds learning rate at peak rate for 90% of the training:
> 
> loss: <img alt="image" width="1196" src="https://user-images.githubusercontent.com/3497946/154183188-acd81628-77e1-4dba-8438-a874e262f8e7.png">
> 
> pred var: <img alt="image" width="1203" src="https://user-images.githubusercontent.com/3497946/154183229-7bb26431-76c6-4c1b-8399-1c312c0bd84b.png">
> 
> target var: <img alt="image" width="1208" src="https://user-images.githubusercontent.com/3497946/154183275-06352bfe-2ff8-44b9-8baf-0811bb8a62c4.png">

@alexeib thank you for the logs. This behavior of the loss (going down and up and down again) is related to the near-optimality in BYOL (Bootstrap Your Own Latent) ?. The BYOL demonstrate the importance of near-optimal predictor for preventing collapse.

**Comment by a43992899** [comment]
> hey, if variances are jumping up and down, that looks like a collapse and you may want to lower you learning rate. i dont have variance plots for the nlp models (wasnt logging it when it was trained) but here is the loss curve:
> 
> <img alt="image" width="1294" src="https://user-images.githubusercontent.com/3497946/154182962-414ea6dd-8264-429e-b357-c69c77aaa69c.png">
> 
> here are a couple example plots from a reduced speech setup with variance etc (it should look somewhat similar with nlp but not exactly the same). this is also using a tri-stage lr scheduler that holds learning rate at peak rate for 90% of the training:
> 
> loss: <img alt="image" width="1196" src="https://user-images.githubusercontent.com/3497946/154183188-acd81628-77e1-4dba-8438-a874e262f8e7.png">
> 
> pred var: <img alt="image" width="1203" src="https://user-images.githubusercontent.com/3497946/154183229-7bb26431-76c6-4c1b-8399-1c312c0bd84b.png">
> 
> target var: <img alt="image" width="1208" src="https://user-images.githubusercontent.com/3497946/154183275-06352bfe-2ff8-44b9-8baf-0811bb8a62c4.png">

@alexeib Hi, why does the NLP curve looks so different from the speech curve, and seems like it does not converge? Actually, my audio modality training curve is similar to your nlp curve. And my predict and target var does not seem to be collapsing. How do I check if my audio model converges?

<img width="1355" alt="image" src="https://user-images.githubusercontent.com/30723625/193879434-2c67f43a-a9bc-48b1-8b91-7a6f1e95f280.png">
<img width="1358" alt="image" src="https://user-images.githubusercontent.com/30723625/193879676-1c08aac5-6a41-4a8a-afeb-9cb3f1c619f8.png">
<img width="1352" alt="image" src="https://user-images.githubusercontent.com/30723625/193879721-3be1464d-2b35-47fc-ae22-9b2b7edcb8b4.png">

**Comment by saurabh-kataria** [comment]
FWIW, I was getting `target_var` < 0.1 error while training data2vec2.0 on my data. I lowered the learning rate from (default) 0.00075 to 0.00050 and the error vanished.

**Comment by orena1** [comment]
Thanks @saurabh-kataria  it did help a bit to reduce the lr, but it still cush with target_val<0.01, did you manage to get more stable solutions?
